Refactors the music player.

GraionDilach opened the pull request:
Proper fix for #8559.
Closes  #8671, #8587 and #8033.

<del>ATM it has one regression: due to the playlist getting set up before the shellmap's lua scripts, the game plays a maximum of two seconds maybe before the shellmap music at game start, depending on settings. That will probably need a StartingMusic trait.</del>

Commit:
Refactors the music player.

Mailaender commented:
This now always play the first song in the playlist, instead of the one specified. https://github.com/OpenRA/OpenRA/blob/bleed/mods/cnc/maps/shellmap/shellmap.lua#L30

> it has one regression: due to the playlist getting set up before the shellmap's lua scripts, the game plays a maximum of two seconds maybe before the shellmap music at game start

I don't see that as a problem. We want to keep that order.

pchote commented:
This is going in the wrong direction by promoting world-state leakage into a feature, instead of a bug.

GraionDilach commented:
@pchote: That sounded as a feature to me. Resolved.

@Mailaender: Such happens only if you don't have a map1.aud (which is **not** in scores.mix) in your game. I guess transit.mix isn't among the files transferred from an online source.

Commit:
Updates all base mods with the MusicPlaylist trait.

Mailaender commented:
I have `map1.aud` and checked again that it works on bleed. Now it plays 2 seconds "Act on Instinct" (first in the playlist) until it switches to `map1.aud` in the cnc mod which is still bogus.

Mailaender commented:
If I disable `Game.Settings.Sound.MapMusic` it still plays, but only "Act on Instinct".

GraionDilach commented:
That's exactly why I was thinking about a StartingMusic trait. Alright.

Mailaender commented on OpenRA.Game/MusicPlaylist.cs:
```diff
@@ -0,0 +1,105 @@
+
+			currentSong = Game.Settings.Sound.Shuffle ? random.First() : music.First();
+
+			Play();
```
```csharp
if (Game.Settings.Sound.MapMusic)
	Play();
```

Mailaender replied:
Or not at all. Why does the playlist decide to play music on it's own?

GraionDilach replied:
MapMusic is the ShellMap / Mission Music checkbox in the UI and even the current bleed tracks silenced music. If I want to disable music overall then the slider is the option.

Mailaender commented:
> That's exactly why I was thinking about a StartingMusic trait.

We removed in favor of Lua. https://github.com/OpenRA/OpenRA/pull/6880

GraionDilach commented:
Lua should not be used for defining the music at map start due to the speed condition. I cannot solve that without a trait.

pchote commented:
My thoughts on the music situation:
```
13:49 <+pchote> the way i'd like to see the music work is that the game will automatically play music if there is no overrides
13:49 <+pchote> lua and/or traits then have the ability to do two things:
13:49 <+pchote> lock the music to a specific track (disabling user choice), and set the suggested track
13:50 <+pchote> so we can then have missions start with a specific track before continuing on as normal (e.g. AOI in GDI01)
13:50 <+pchote> and we can also have missions and MP force-play the score music at the end of the game
13:51 <+pchote> we could possibly drop the "suggestion" lua API by having it defined in the map.yaml instead
13:51 <+pchote> like we do for videos
```

GraionDilach commented:
Rebased. ATM only the win-lose song is unimplemented but I'd do it in a followup PR.

Mailaender commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,139 @@
+
+	public class MusicPlaylist
+	{
+		readonly MusicPlaylistInfo info;
```
> The private field `OpenRA.Traits.MusicPlaylist.info` is assigned but its value is never used

GraionDilach replied:
Right. It'll be probably used in the followup PRs, but for now let's fix Travis.

GraionDilach commented:
Was the Requested changes label on this PR stayed for the score music implementation?

Mailaender commented:
Please remove the deleted shellmap.lua from https://github.com/GraionDilach/OpenRA/blob/music-refactor/OpenRA.sln, too. Otherwise it seems to retain the behavior now. :white_check_mark:

GraionDilach commented:
Done. Also did the TODO: in the playlist widget and renamed the Loop to StartingMusicLoop to prevent confusion.

penev92 commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,135 @@
+	public class MusicPlaylistInfo : ITraitInfo
+	{
+		public readonly string StartingMusic = null;
+		public readonly bool StartingMusicLoop = false;
```
Let's go with `LoopStartingMusic`.

GraionDilach replied:
Went with that then.

GraionDilach commented on OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs:
```diff
@@ -111,11 +106,10 @@ public MusicPlayerLogic(Widget widget, Ruleset modRules, World world, Action onE
 			currentSong = Sound.CurrentMusic;
-			if (currentSong == null && music.Any())
-				currentSong = Game.Settings.Sound.Shuffle ? random.First() : music.First();
+			if (currentSong == null && musicPlaylist.GetInstalled)
```
Oh wait, this is a bug.

penev92 commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,135 @@
+				.Skip(1).FirstOrDefault() ?? songs.Reverse().FirstOrDefault();
+		}
+
+		public void Dispose()
```
I believe you actually want `public void Disposing(Actor self)` from the `INotifyActorDisposing` interface here. Otherwise this never gets called.
Another issue here is that the new world is initialized (and plays its own music) before the old one is disposed of, meaning this here will never do anything, because by the time it is called, the music has already changed to play whatever the new world specified.

GraionDilach replied:
Since I don't have to do a check on the music now, I refactored this into a StopMusic call, which I call explicitly in World.Dispose - this StopMusic is also the StopMusic Lua and GUI will use.

penev92 replied:
The `World.Dispose()` method already notifies all relevant traits to stop what they are doing and clean up, so I'd rather we go with that instead of hacking a call to `Playlist` inside `World.Dispose()`. (The current `Sound.StopMusic()` in there was a quick hack to patch things up.)

GraionDilach replied:
That would be cool having the fadeout, but the fadeout cannot be in a Dispose... I'd like to evade that discussion for now.

penev92 replied:
I'm not talking about the fadeout. I'm talking about `Playlist.Stop();` being out of place inside `World.Dispose()`.
And we even have an interface for that - `INotifyActorDisposing`, or whatever it was called.

penev92 replied:
P.S.: That actually removes any references from `World` to `Playlist`, which is a very good thing. Not only it spares us some bogus (from an architectural standpoint) class/trait/whateveryouwant dependencies, it lets people change Playlist with their own trait in their mods, as there is no longer a hardcoded reference.

GraionDilach replied:
Except that World does a Trait check on MusicPlaylist and not TraitOrDefault, or should I change that as well?

GraionDilach replied:
PS.: TBH, after I'm done, I'll see no reason why would someone go with a new trait instead of MusicPlaylist because I might implement all the footwork of a faction-specific dynamic music system - the dynamic part is the one I'm wondering about for now, but whatever.

GraionDilach replied:
PS2.: Added the interface implementation.

penev92 commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,135 @@
+				repeat = Game.Settings.Sound.Repeat;
+			}
+
+			Play();
```
May I suggest
```
			// Fade out any previous music before we replace it with new one
			var volume = Sound.MusicVolume;

			while (Sound.MusicVolume > 0.0f)
				Sound.MusicVolume -= 0.0000005f;

			Sound.MusicVolume = volume;
```
before this line.

GraionDilach replied:
There's only one stream for the music. I'd guess that would do a fade-out within a single tick. Which makes no difference for the ear.

penev92 replied:
No, it actually may prolong the tick, but it sounds quite nice. ( I did test it )

penev92 commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,135 @@
+		MusicInfo[] random;
+
+		public MusicInfo[] GetPlaylist { get { return music; } }
+		public bool GetInstalled { get { return installed; } }
```
How about simplifying things to
```
		public MusicInfo[] Playlist { get; private set; }
		public bool IsInstalled { get; private set; }
```

penev92 commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,135 @@
+
+		MusicInfo currentSong = null;
+		MusicInfo[] music;
+		MusicInfo[] random;
```
Can be `readonly`.

penev92 commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,135 @@
+		bool installed = false;
+		bool repeat;
+
+		MusicInfo currentSong = null;
```
Redundant default value.

penev92 commented:
Depends on #8607 now for proper disposing.

GraionDilach commented:
Rebased, addressed most issues (excluding fadeout which will be a debate, I feel), removed a pre-#8560 workaround to deal with leaking music (shellmap-to-modchooser).

**Due to #8607 fixing world leaking and this making music world-dependant, we lost loadscreen music.**

I am not certain if I should call that bug or feature, because while the default mods have small loadtimes, nothing stops a mod to grow enough large for a noticable silence.

pchote commented on OpenRA.Game/World.cs:
```diff
@@ -169,6 +169,8 @@ public bool ToggleInputMode<T>() where T : IOrderGenerator, new()
 			}
 		}
 
+		public readonly MusicPlaylist Playlist;
```
This is only accessed from two places (`MediaGlobal` and `MusicPlayerLogic`).  It would be better if you queried the trait in those two places and removed this reference.

GraionDilach replied:
The World does not implement the self.Trait<T>() interface.

pchote replied:
In the `MediaGlobal` constructor add `playlist = world.WorldActor.Trait<MusicPlaylist>()` and a `readonly MusicPlaylist playlist;` instance variable.  You can do something similar in `MusicPlayerLogic`.

GraionDilach replied:
Thanks, for some reason IntelliSense hid that option at first. Done.

pchote commented on OpenRA.Mods.Common/Scripting/Global/MediaGlobal.cs:
```diff
@@ -27,10 +26,12 @@ namespace OpenRA.Mods.Common.Scripting
 	public class MediaGlobal : ScriptGlobal
 	{
 		World world;
+		MusicPlaylist playlist;
```
This and the world above should be readonly, and then a newline before the constructor please.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs:
```diff
@@ -18,29 +18,24 @@ namespace OpenRA.Mods.Common.Widgets.Logic
-		MusicInfo[] music;
-		MusicInfo[] random;
-		ScrollPanelWidget musicList;
+		MusicPlaylist musicPlaylist;
```
Readonly here and for the two widgets below please.  The general rule is that if it's set in a constructor and isn't supposed to change, then it should be marked readonly.

GraionDilach replied:
Done.

pchote commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,150 @@
+		readonly MusicInfo[] playlist;
+
+		MusicInfo currentSong;
+		bool isMusicInstalled;
```
This can be readonly and moved below `playlist`.

pchote replied:
You can then make this public and remove the `IsInstalled()` method.

GraionDilach replied:
Done.

pchote commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,150 @@
+
+	public class MusicPlaylist : INotifyActorDisposing
+	{
+		readonly MusicInfo[] random;
```
These can become public too, and then remove the methods that return them.

GraionDilach replied:
Outer classes don't need the random - only the music player GUI uses playlist alone to list the available songs - and I plan to filter the songs in the return method in a followup PR aimed for faction-specific OST.

pchote replied:
Yeah, I noticed that after commenting.  It still applies to `playlist`, though.

GraionDilach replied:
Added one above the return parameter.

pchote commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,150 @@
+
+				if (!repeat)
+					currentSong = GetNextSong();
+				Play();
```
nit: newline above here please.

pchote commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,150 @@
+
+				if (!repeat)
+					currentSong = GetNextSong();
+				Play();
```
nit: newline above here please.

pchote commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,150 @@
+				.Skip(1).FirstOrDefault() ?? songs.FirstOrDefault();
+		}
+
+		public MusicInfo GetPrevSong()
```
You could remove a bit of duplication by having a private `GetNextSong(IEnumerable<MusicInfo> tracks)` that gets called with random/playlist from `GetNextSong` and (random/playlist).Reverse() from `GetPrevSong`.

GraionDilach replied:
I got confused in the syntax, so I just made the two public functions into wrappers of a private one instead.

pchote commented on OpenRA.Mods.Common/Scripting/Global/MediaGlobal.cs:
```diff
@@ -53,15 +54,11 @@ public void PlayMusic(string track = null, LuaFunction func = null)
-			: Game.Settings.Sound.Shuffle ? music.Random(Game.CosmeticRandom)
-			: previousMusic == null ? music.First()
-			: music.SkipWhile(s => s != previousMusic).Skip(1).First();
+			: playlist.GetNextSong();
```
Please throw an error if someone asks for a bogus track instead of silently falling back to a random song.

pchote commented on mods/cnc/maps/shellmap/map.yaml:
```diff
@@ -994,6 +994,9 @@ Rules:
 		LuaScript:
 			Scripts: shellmap.lua
+		MusicPlaylist:
+			StartingMusic: map1
```
You're leaking state somewhere.  If you start TD without scores.mix it will play map1.  Go Extras &rarr; Music &rarr; Install Music &rarr; Back and it will sometimes play the score screen music.

pchote replied:
Actually, it seems to be ignoring this completely, and picking a random track at startup.

GraionDilach replied:
Yeah, https://github.com/OpenRA/OpenRA/pull/8562/files#diff-0612b1f741b8612131b70560bfabc77aR57 is a check I carried over which causes it. I guess it should be removed there.

pchote commented:
Looks good overall, but there are a few style nits and you'll need to fix whatever is stopping `StartingMusic` from working on the TD shellmap.

GraionDilach commented:
On a second check, noidea what caused the bug in the TD shellmap, but I got it working now. Also addressed the nits.

GraionDilach commented:
Note: This will fix #8033 - I was completely unaware that was reported.

penev92 commented:
Now that #8528 was merged, you'll have to account for a few more missions with the music changes, sorry.

GraionDilach commented:
No problem, I actually hoped for this result - didn't wanted to get clemty into trouble of a rebase.

GraionDilach commented:
Rebased. I think no changes are requested at the moment - pchote's changes were already resolved before I went out during the weekend.

obrakmann commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,149 @@
+				repeat = Game.Settings.Sound.Repeat;
+			}
+
+			Play();
```
This isn't correct, it completely ignores `Game.Settings.Sounds.MapMusic` (try starting the game with music enabled, then disable the checkbox, then switch mods). You'd need to do it like this:

```csharp
if (!Game.Settings.Sound.MapMusic)
    return;

if (!string.IsNullOrEmpty(info.StartingMusic) && ...)
    ...
else
    ...

Play()
```

pchote replied:
Why don't we just remove MapMusic?  If people don't want music, then they can mute it.

obrakmann replied:
Not sure if I'd like that. The game would constantly do something without any benefit. That feels unclean.

pchote replied:
The setting was added long ago to work around the type of bugs that this pr is finally fixing properly.  It's mostly redundant now, and fouls up the logic around end-of-game scripted music.

GraionDilach replied:
@obrakmann I fail to see what's wrong with that. I interpreted MapMusic as the allow override of the music playlist. The sole thing which fails is to carry over the active song, but due to music being part of the world, I don't think that's a bug.

OTOH decide if you still want that checkbox afterall.

obrakmann replied:
Really not sure about all this, but I'll defer the decision to you guys, as I clearly have the minority opinion here. I'd probably vote for removing the checkbox then.

GraionDilach replied:
Removing the checkbox sounds like the general agreement - even I was considering that after this. Done.

obrakmann commented on OpenRA.sln:
```diff
@@ -66,7 +66,6 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Red Alert Lua scripts", "Re
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Dune 2000 Lua scripts", "Dune 2000 Lua scripts", "{06B1AE07-DDB0-4287-8700-A8CD9A0E652E}"
 	ProjectSection(SolutionItems) = preProject
-		mods\d2k\maps\shellmap\shellmap.lua = mods\d2k\maps\shellmap\shellmap.lua
```
That looks like an accidental removal.

penev92 replied:
Iirc the only thing that script did was play the music, which is now done by the new trait, so the scrip is now redundant.

GraionDilach replied:
This might need a revert anyway due to the scripted lighting pr.

GraionDilach replied:
Yes, I reverted removing the file from the solution to not get a conflict with #8684.

abcdefg30 commented on OpenRA.Game/Traits/World/MusicPlaylist.cs:
```diff
@@ -0,0 +1,149 @@
+			IsMusicAvailable = world.Map.Rules.InstalledMusic.Any();
+
+			if (!IsMusicAvailable)
+			{
```
I think you can drop the braces here.

GraionDilach replied:
Dropped.

Commit:
Removes the MapMusic checkbox.

GraionDilach commented:
Also did a bugfix to fall back to available music from the playlist if the music selected in trait is defined in the yamls, but missing from the user PC (TD missions without music pack would not play silence anymore).

obrakmann commented:
Just got this in d2k:
```
Exception of type `System.Reflection.TargetInvocationException`: Exception has been thrown by the target of an invocation.
Inner Exception of type `System.ArgumentNullException`: Argument cannot be null.
Parameter name: source
    at System.Linq.Check.Source (System.Object source) [0x00000] in <filename unknown>:0 
  at System.Linq.Enumerable.Any[MusicInfo] (IEnumerable`1 source) [0x00000] in <filename unknown>:0 
  at OpenRA.Mods.Common.Widgets.Logic.MusicPlayerLogic.BuildMusicTable () [0x0002a] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs:111 
  at OpenRA.Mods.Common.Widgets.Logic.MusicPlayerLogic..ctor (OpenRA.Widgets.Widget widget, OpenRA.Ruleset modRules, OpenRA.World world, System.Action onExit) [0x0007b] in /home/oliver/devel/openra/git/OpenRA.Mods.Common/Widgets/Logic/MusicPlayerLogic.cs:36 
```

GraionDilach commented:
Whoops, fixed.

obrakmann commented:
Thanks. :+1: when the last two (or at least the last) commits are squashed into the first.

GraionDilach commented:
Squashed the last commit. I wouldn't squash the MapMusic one to make sure that change doesn't get overlooked - it just loosely fits to this PR anyway.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/128f27f87edf1803baa45ae7cb7ff05ab24746dc%5E...128f27f87edf1803baa45ae7cb7ff05ab24746dc)
